File: editors/jsfcstm/src/diagnostics/analyzers/design-health.ts
import type {
    ActionInfo,
    EventInfo,
    ForcedTransitionInfo,
    ModelDiagnosticJson,
    ModelMetrics,
    ModelSpanJson,
    StateInfo,
    TransitionInfo,
    VariableInfo,
} from '../inspect';
import type {StateMachine} from '../../model/runtime';
import {refsWithSuggestedFix} from '../suggested-fix';
import {collectComboWarnings} from './combo';
import {collectConstFoldWarnings} from './const-fold';
import {collectDataFlowWarnings} from './data-flow';
import {collectNamingWarnings} from './naming';
import {collectNumericWarnings} from './numeric';
import {collectRedundancyWarnings} from './redundancy';
import {collectStructuralWarnings} from './structural';
import {collectThresholdWarnings, type ThresholdOptions} from './thresholds';
import {collectTransitionInfos} from './transition-info';
import {collectTypeWarnings} from './type-shape';

export function collectDesignHealthWarnings(
    states: StateInfo[],
    transitions: TransitionInfo[],
    variables: VariableInfo[],
    events: EventInfo[],
    actions: ActionInfo[],
    forcedTransitions: ForcedTransitionInfo[],
    metrics: ModelMetrics,
    reachabilityGraph: Record<string, string[]>,
    rootStatePath?: string,
    thresholds?: ThresholdOptions,
    machine?: StateMachine,
): ModelDiagnosticJson[] {
    const thresholdOptions = thresholds ?? {
        deepHierarchyThreshold: 6,
        largeCompositeThreshold: 12,
        varToLeafRatioThreshold: 2.0,
    };
    const diagnostics = [
        ...collectUnreachableStateDiagnostics(states, reachabilityGraph, rootStatePath),
        ...collectUnreachableTransitionDiagnostics(states, transitions, reachabilityGraph, rootStatePath),
        ...(machine ? collectConstFoldWarnings(machine) : collectGuardConstFalseDiagnostics(transitions)),
        ...(machine ? collectComboWarnings(machine) : []),
        ...collectUnusedEventDiagnostics(events),
        ...collectStructuralWarnings(
            states,
            transitions,
            actions,
            forcedTransitions,
            reachabilityGraph,
            rootStatePath,
        ),
        ...collectThresholdWarnings(states, metrics, thresholdOptions),
        ...collectNamingWarnings(actions),
        ...collectTypeWarnings(variables),
        ...collectDataFlowWarnings(variables, machine),
        ...(machine ? collectNumericWarnings(machine) : []),
        ...collectRedundancyWarnings(transitions, events, states),
        ...collectTransitionInfos(states, transitions),
    ];
    return diagnostics.map(diag => ({
        ...diag,
        refs: refsWithSuggestedFix(diag.code, diag.refs),
    }));
}

function collectUnreachableStateDiagnostics(
    states: StateInfo[],
    reachabilityGraph: Record<string, string[]>,
    rootStatePath?: string,
): ModelDiagnosticJson[] {
    if (states.length === 0) return [];
    const rootPath = rootStatePath ?? states[0].path;
    const reachable = new Set<string>(reachabilityGraph[rootPath] ?? []);
    reachable.add(rootPath);
    const out: ModelDiagnosticJson[] = [];
    for (const state of states) {
        if (state.is_pseudo || reachable.has(state.path)) continue;
        out.push({
            code: 'W_UNREACHABLE_STATE',
            severity: 'warning',
            message: `State ${JSON.stringify(state.path)} is unreachable from the root entry path.`,
            span: null,
            refs: {state_path: state.path},
        });
    }
    return out;
}

function collectUnreachableTransitionDiagnostics(
    states: StateInfo[],
    transitions: TransitionInfo[],
    reachabilityGraph: Record<string, string[]>,
    rootStatePath?: string,
): ModelDiagnosticJson[] {
    if (states.length === 0) return [];
    const rootPath = rootStatePath ?? states[0].path;
    const statePaths = new Set(states.map(state => state.path));
    const reachable = new Set<string>(reachabilityGraph[rootPath] ?? []);
    reachable.add(rootPath);
    const emittedComboOrigins = new Set<string>();
    const comboEndpoints = buildComboOriginEndpoints(transitions);
    const out: ModelDiagnosticJson[] = [];

    for (const transition of transitions) {
        const comboOriginIds = Array.from(new Set(
            transition.combo_origin_refs
                .map(ref => ref.origin_id)
                .filter((originId): originId is string => typeof originId === 'string'),
        )).sort();
        if (comboOriginIds.length > 0) {
            for (const originId of comboOriginIds) {
                if (emittedComboOrigins.has(originId)) continue;
                emittedComboOrigins.add(originId);
                const originRef = transition.combo_origin_refs.find(ref => ref.origin_id === originId);
                const endpoint = comboEndpoints.get(originId);
                if (!originRef || !endpoint?.targetPath) continue;
                const lookupPath = endpoint.selectionOwnerPath ?? endpoint.sourcePath;
                if (!lookupPath || reachable.has(lookupPath)) continue;
                out.push({
                    code: 'W_UNREACHABLE_TRANSITION',
                    severity: 'warning',
                    message: `Transition ${JSON.stringify(endpoint.sourcePath ?? '[*]')} -> ${JSON.stringify(endpoint.targetPath)} has a source outside the guard-agnostic root-reachable topology.`,
                    span: originRef?.transition_span ?? null,
                    refs: {
                        reason: 'source_unreachable',
                        verification_scope: 'topological_only',
                        from_path: endpoint.sourcePath ?? '[*]',
                        to_path: endpoint.targetPath,
                        source_state_path: endpoint.sourcePath,
                        selection_owner_path: endpoint.selectionOwnerPath,
                        transition_index: null,
                        forced_origin: null,
                        combo_origin_ids: [originId],
                    },
                });
            }
            continue;
        }

        const selection = transitionSelectionPaths(transition, statePaths);
        if (!selection.ownerPath || !statePaths.has(selection.ownerPath)) continue;
        if (reachable.has(selection.ownerPath)) continue;

        const message = transition.is_forced
            ? `Generated forced transition expansion ${JSON.stringify(transition.from_path)} -> ${JSON.stringify(transition.to_path)} has a source outside the guard-agnostic root-reachable topology.`
            : `Transition ${JSON.stringify(transition.from_path)} -> ${JSON.stringify(transition.to_path)} has a source outside the guard-agnostic root-reachable topology.`;
        out.push({
            code: 'W_UNREACHABLE_TRANSITION',
            severity: 'warning',
            message,
            span: transitionSourceSpan(transition),
            refs: {
                reason: 'source_unreachable',
                verification_scope: 'topological_only',
                from_path: transition.from_path,
                to_path: transition.to_path,
                source_state_path: selection.sourcePath,
                selection_owner_path: selection.sourcePath === null ? selection.ownerPath : null,
                transition_index: transition.transition_index,
                forced_origin: transition.forced_origin,
                combo_origin_ids: comboOriginIds,
            },
        });
    }
    return out;
}

interface ComboOriginEndpoints {
    sourcePath: string | null;
    selectionOwnerPath: string | null;
    targetPath: string | null;
}

function buildComboOriginEndpoints(
    transitions: TransitionInfo[],
): Map<string, ComboOriginEndpoints> {
    const endpoints = new Map<string, ComboOriginEndpoints>();
    for (const transition of transitions) {
        const projection = comboProjectionEndpoints(transition);
        for (const ref of transition.combo_origin_refs) {
            const current = endpoints.get(ref.origin_id) ?? {
                sourcePath: null,
                selectionOwnerPath: null,
                targetPath: null,
            };
            current.sourcePath = ref.source_path ?? projection.sourcePath ?? current.sourcePath;
            current.selectionOwnerPath = ref.selection_owner_path
                ?? projection.selectionOwnerPath
                ?? current.selectionOwnerPath;
            if (ref.target_path !== null) {
                current.targetPath = ref.target_path;
            } else if (ref.role === 'terminal') {
                current.targetPath = transition.to_path;
            }
            endpoints.set(ref.origin_id, current);
        }
    }
    return endpoints;
}

function comboProjectionEndpoints(
    transition: TransitionInfo,
): ComboOriginEndpoints {
    const projection = transition.combo_projection_key;
    if (!Array.isArray(projection) || !Array.isArray(projection[0])) {
        return {
            sourcePath: transition.from_path === '[*]' ? null : transition.from_path,
            selectionOwnerPath: null,
            targetPath: null,
        };
    }
    const ownerPath = projection[0].every(item => typeof item === 'string')
        ? (projection[0] as string[]).join('.')
        : null;
    if (projection[1] === 'entry') {
        return {sourcePath: null, selectionOwnerPath: ownerPath, targetPath: null};
    }
    const sourcePath = Array.isArray(projection[2])
        && projection[2].every(item => typeof item === 'string')
        ? (projection[2] as string[]).join('.')
        : null;
    return {sourcePath, selectionOwnerPath: null, targetPath: null};
}

function transitionSelectionPaths(
    transition: TransitionInfo,
    statePaths: Set<string>,
): {sourcePath: string | null; ownerPath: string | null} {
    if (transition.from_path !== '[*]') {
        return {sourcePath: transition.from_path, ownerPath: transition.from_path};
    }
    if (transition.to_path === '[*]' || !transition.to_path.includes('.')) {
        return {sourcePath: null, ownerPath: null};
    }
    const ownerPath = transition.to_path.slice(0, transition.to_path.lastIndexOf('.'));
    if (!statePaths.has(ownerPath)) return {sourcePath: null, ownerPath: null};
    return {sourcePath: null, ownerPath};
}

function transitionSourceSpan(transition: TransitionInfo): ModelSpanJson | null {
    const range = transition.__sourceRange;
    if (!range) return null;
    return {
        line: range.start.line + 1,
        column: range.start.character + 1,
        end_line: range.end.line + 1,
        end_column: range.end.character + 1,
    };
}

function collectGuardConstFalseDiagnostics(transitions: TransitionInfo[]): ModelDiagnosticJson[] {
    const out: ModelDiagnosticJson[] = [];
    for (const transition of transitions) {
        if (!isMinimalConstFalseGuard(transition.guard)) continue;
        out.push({
            code: 'W_GUARD_CONST_FALSE',
            severity: 'warning',
            message: `Transition ${JSON.stringify(transition.from_path)} -> ${JSON.stringify(transition.to_path)} has a guard that is statically false.`,
            span: null,
            refs: {
                transition_span: null,
                folded_value: false,
                from_path: transition.from_path,
                to_path: transition.to_path,
                guard_text: transition.guard,
                transition_index: transition.transition_index,
            },
        });
    }
    return out;
}

function isMinimalConstFalseGuard(guard: string | null): boolean {
    if (guard === null) return false;
    const normalized = guard.trim().toLowerCase();
    if (normalized === 'false' || normalized === '0') return true;
    const match = /^\s*([+-]?\d+)\s*==\s*([+-]?\d+)\s*$/.exec(guard);
    return match !== null && normalizeDecimalInteger(match[1]) !== normalizeDecimalInteger(match[2]);
}

function normalizeDecimalInteger(value: string): string {
    let sign = '';
    let digits = value;
    if (digits.startsWith('+') || digits.startsWith('-')) {
        sign = digits[0] === '-' ? '-' : '';
        digits = digits.slice(1);
    }
    digits = digits.replace(/^0+/, '');
    if (digits === '') return '0';
    return sign + digits;
}

function collectUnusedEventDiagnostics(events: EventInfo[]): ModelDiagnosticJson[] {
    const out: ModelDiagnosticJson[] = [];
    for (const event of events) {
        if (!event.is_declared || event.is_used) continue;
        out.push({
            code: 'W_UNUSED_EVENT',
            severity: 'warning',
            message: `Event ${JSON.stringify(event.qualified_name)} is declared but never used.`,
            span: null,
            refs: {
                event_qualified_name: event.qualified_name,
                scope: event.scope,
            },
        });
    }
    return out;
}

File: pyfcstm/diagnostics/analyzers/design_health.py
"""Design-health diagnostics derived from inspect-surface data.

The reachability helper keeps the root itself in the result, even when the
inspect graph only lists outgoing paths:

    >>> sorted(_reachable_state_paths({'Root': ('Root.Active',)}, 'Root'))
    ['Root', 'Root.Active']
"""

import re
from dataclasses import replace
from typing import TYPE_CHECKING, Iterable, List, Optional

from ..suggested_fix import refs_with_suggested_fix
from ...utils.validate import ModelDiagnostic
from .combo import collect_combo_warnings
from .const_fold import collect_const_fold_warnings
from .data_flow import collect_data_flow_warnings
from .naming import collect_naming_warnings
from .numeric import collect_numeric_warnings
from .redundancy import collect_redundancy_warnings
from .structural import collect_structural_warnings
from .thresholds import collect_threshold_warnings
from .transition_info import collect_transition_infos
from .type_shape import collect_type_warnings

if TYPE_CHECKING:  # pragma: no cover - import-time type hints only
    from ..inspect import (
        ActionInfo,
        EventInfo,
        ForcedTransitionInfo,
        ModelMetrics,
        StateInfo,
        TransitionInfo,
        VariableInfo,
    )
    from ...model.model import StateMachine


def collect_design_health_warnings(
    states: Iterable['StateInfo'],
    transitions: Iterable['TransitionInfo'],
    variables: Iterable['VariableInfo'],
    events: Iterable['EventInfo'],
    actions: Iterable['ActionInfo'],
    forced_transitions: Iterable['ForcedTransitionInfo'],
    metrics: 'ModelMetrics',
    reachability_graph,
    root_state_path: Optional[str] = None,
    deep_hierarchy_threshold: int = 6,
    large_composite_threshold: int = 12,
    var_to_leaf_ratio_threshold: float = 2.0,
    machine: Optional['StateMachine'] = None,
) -> List[ModelDiagnostic]:
    """Collect design-health warning diagnostics from inspect payloads."""
    diagnostics: List[ModelDiagnostic] = []
    states = list(states)
    transitions = list(transitions)
    variables = list(variables)
    events = list(events)
    actions = list(actions)
    forced_transitions = list(forced_transitions)
    resolved_root_state_path = _resolve_root_state_path(states, root_state_path)
    diagnostics.extend(_unreachable_state_diagnostics(
        states,
        reachability_graph,
        resolved_root_state_path,
    ))
    diagnostics.extend(_unreachable_transition_diagnostics(
        states,
        transitions,
        reachability_graph,
        resolved_root_state_path,
    ))
    diagnostics.extend(
        collect_const_fold_warnings(machine)
        if machine is not None
        else _guard_const_false_diagnostics(transitions)
    )
    diagnostics.extend(collect_combo_warnings(machine))
    diagnostics.extend(_unused_event_diagnostics(events))
    diagnostics.extend(collect_structural_warnings(
        states,
        transitions,
        actions,
        forced_transitions,
        reachability_graph,
        root_state_path=resolved_root_state_path,
    ))
    diagnostics.extend(collect_threshold_warnings(
        states,
        metrics,
        deep_hierarchy_threshold=deep_hierarchy_threshold,
        large_composite_threshold=large_composite_threshold,
        var_to_leaf_ratio_threshold=var_to_leaf_ratio_threshold,
    ))
    diagnostics.extend(collect_naming_warnings(actions))
    diagnostics.extend(collect_type_warnings(variables))
    diagnostics.extend(collect_data_flow_warnings(variables, machine))
    diagnostics.extend(collect_redundancy_warnings(transitions, events, states))
    diagnostics.extend(collect_transition_infos(states, transitions))
    diagnostics.extend(collect_numeric_warnings(machine))
    return _with_suggested_fixes(diagnostics)


def _with_suggested_fixes(diagnostics: List[ModelDiagnostic]) -> List[ModelDiagnostic]:
    return [
        replace(diag, refs=refs_with_suggested_fix(diag.code, diag.refs))
        for diag in diagnostics
    ]


def _resolve_root_state_path(states, root_state_path):
    if root_state_path is not None:
        return root_state_path
    if not states:
        return None
    return states[0].path


def _unreachable_state_diagnostics(states, reachability_graph, root_state_path) -> List[ModelDiagnostic]:
    if not states or root_state_path is None:
        return []
    reachable = _reachable_state_paths(reachability_graph, root_state_path)
    diagnostics: List[ModelDiagnostic] = []
    for state in states:
        if not state.is_leaf or state.is_pseudo or state.path in reachable:
            continue
        diagnostics.append(
            ModelDiagnostic(
                code='W_UNREACHABLE_STATE',
                severity='warning',
                message=f'State {state.path!r} is unreachable from the root entry path.',
                span=state.span,
                refs={'state_path': state.path},
            )
        )
    return diagnostics


def _reachable_state_paths(reachability_graph, root_state_path):
    reachable = set(reachability_graph.get(root_state_path, ()))
    reachable.add(root_state_path)
    return reachable


def _unreachable_transition_diagnostics(
        states,
        transitions,
        reachability_graph,
        root_state_path,
) -> List[ModelDiagnostic]:
    """Report authored transitions outside the guard-agnostic topology.

    A normal transition is selected from its source state.  An initial
    transition is selected by the composite that owns its ``[*]`` entry; its
    owner is recoverable from the direct target path in the inspect payload.
    Combo relay edges are projected to one warning per authored combo origin so
    generated edges do not multiply one source finding.
    """
    if not states or root_state_path is None:
        return []
    state_paths = {state.path for state in states}
    reachable = _reachable_state_paths(reachability_graph, root_state_path)
    diagnostics: List[ModelDiagnostic] = []
    emitted_combo_origins = set()
    for transition in transitions:
        combo_origin_ids = tuple(sorted({
            ref.origin_id
            for ref in transition.combo_origin_refs
            if isinstance(ref.origin_id, str)
        }))
        if combo_origin_ids:
            for origin_id in combo_origin_ids:
                if origin_id in emitted_combo_origins:
                    continue
                emitted_combo_origins.add(origin_id)
                origin_ref = next(
                    (
                        ref for ref in transition.combo_origin_refs
                        if ref.origin_id == origin_id
                    ),
                    None,
                )
                if origin_ref is None or origin_ref.target_path is None:
                    continue
                source_state_path = origin_ref.source_path
                selection_owner_path = origin_ref.selection_owner_path
                lookup_path = selection_owner_path or source_state_path
                if lookup_path is None:
                    continue
                if lookup_path in reachable:
                    continue
                diagnostics.append(ModelDiagnostic(
                    code='W_UNREACHABLE_TRANSITION',
                    severity='warning',
                    message=(
                        f'Transition {origin_id!r} has a source outside the '
                        'guard-agnostic root-reachable topology.'
                    ),
                    span=transition.span if origin_ref is None else origin_ref.transition_span,
                    refs={
                        'reason': 'source_unreachable',
                        'verification_scope': 'topological_only',
                        'from_path': '[*]' if source_state_path is None else source_state_path,
                        'to_path': origin_ref.target_path,
                        'source_state_path': source_state_path,
                        'selection_owner_path': selection_owner_path,
                        'transition_index': None,
                        'forced_origin': None,
                        'combo_origin_ids': [origin_id],
                    },
                ))
            continue

        source_state_path, selection_owner_path = _transition_selection_paths(
            transition,
            state_paths,
        )
        if selection_owner_path is None or selection_owner_path not in state_paths:
            continue
        if selection_owner_path in reachable:
            continue

        forced_expansion = transition.is_forced
        diagnostics.append(ModelDiagnostic(
            code='W_UNREACHABLE_TRANSITION',
            severity='warning',
            message=(
                (
                    f'Generated forced transition expansion '
                    f'{transition.from_path!r} -> {transition.to_path!r} '
                    'has a source outside the guard-agnostic root-reachable topology.'
                )
                if forced_expansion else (
                    f'Transition {transition.from_path!r} -> {transition.to_path!r} '
                    'has a source outside the guard-agnostic root-reachable topology.'
                )
            ),
            span=transition.span,
            refs={
                'reason': 'source_unreachable',
                'verification_scope': 'topological_only',
                'from_path': transition.from_path,
                'to_path': transition.to_path,
                'source_state_path': source_state_path,
                'selection_owner_path': selection_owner_path if transition.from_path == '[*]' else None,
                'transition_index': transition.transition_index,
                'forced_origin': transition.forced_origin,
                'combo_origin_ids': list(combo_origin_ids),
            },
        ))
    return diagnostics


def _transition_selection_paths(transition, state_paths):
    if transition.from_path != '[*]':
        return transition.from_path, transition.from_path
    if transition.to_path == '[*]' or '.' not in transition.to_path:
        return None, None
    owner_path = transition.to_path.rsplit('.', 1)[0]
    if owner_path not in state_paths:
        return None, None
    return None, owner_path


def _guard_const_false_diagnostics(transitions) -> List[ModelDiagnostic]:
    diagnostics: List[ModelDiagnostic] = []
    for transition in transitions:
        if _is_minimal_const_false_guard(transition.guard):
            diagnostics.append(
                ModelDiagnostic(
                    code='W_GUARD_CONST_FALSE',
                    severity='warning',
                    message=(
                        f'Transition {transition.from_path!r} -> {transition.to_path!r} '
                        'has a guard that is statically false.'
                    ),
                    span=transition.span,
                    refs={
                        'transition_span': transition.span,
                        'folded_value': False,
                        'from_path': transition.from_path,
                        'to_path': transition.to_path,
                        'guard_text': transition.guard,
                        'transition_index': transition.transition_index,
                    },
                )
            )
    return diagnostics


def _is_minimal_const_false_guard(guard_text) -> bool:
    if guard_text is None:
        return False
    normalized = guard_text.strip().lower()
    if normalized in {'false', '0'}:
        return True
    match = re.match(r'^\s*([+-]?\d+)\s*==\s*([+-]?\d+)\s*$', guard_text)
    if match is not None:
        return int(match.group(1)) != int(match.group(2))
    return False


def _unused_event_diagnostics(events) -> List[ModelDiagnostic]:
    diagnostics: List[ModelDiagnostic] = []
    for event in events:
        if not event.is_declared or event.is_used:
            continue
        diagnostics.append(
            ModelDiagnostic(
                code='W_UNUSED_EVENT',
                severity='warning',
                message=f'Event {event.qualified_name!r} is declared but never used.',
                span=event.span,
                refs={
                    'event_qualified_name': event.qualified_name,
                    'scope': event.scope,
                },
            )
        )
    return diagnostics
